### PR TITLE
[XLA] Enable vectorization of row broadcasting

### PR DESCRIPTION
**Pull request by nouiz**
Reland https://github.com/tensorflow/tensorflow/pull/48381 with a fix

Mostly the refactoring done in the commit https://github.com/tensorflow/tensorflow/pull/48381/commits/247b1d3bc60d80662541a0aff7423025cbe8ba05

introduced the bug. Sometimes, the new simpler indexing was triggered even when we didn't try to vectorize the row broadcasting. The inputs of the fusion wasn't c contiguous in this case.

Now, I pass directly if we enable the row vectorization or not. So the simpler indexing can only be triggered when we try to vectorize row broadcasting.

@cheshire

### CODE REVIEW

**Comment by cheshire** [comment]
Do we have the regression test case?

**Comment by nouiz** [comment]
> Do we have the regression test case?

I have it. But it is ~40MB due to the big contant. So I can't reuse it.
I was able to make one test, but it wasn't trivial. I amended the last commit to include it.

I had 2 main problems:
- I changed the constant to be parameter inputs but this doesn't reproduce the bug.
- I tried to use a FileCheck to check the indexing, but I have this error:
```
%tmp_12 = f32[32,7,7,352]{3,2,1,0} convolution(f32[32,7,7,1024]{3,2,1,0} %tmp_10, f32[1,1,1024,352]{3,2,1,0} %tmp_11), window={size=1x1}, dim_labels=b01f_01io->b01f2021-04-15 15:40:54.001613: F tensorflow/compiler/xla/service/gpu/tests/hlo_to_llvm_ir.cc:135] Non-OK-status: CompileAndPrintLlvmIrFromFile(argv[1], ptx, sm) status: Internal: LHLO opcode convolution is not supported.
```

So finally, I added the optimized graph, but remove the convolution.

**Review by cheshire — APPROVED** [review]

File: tensorflow/compiler/xla/service/gpu/parallel_loop_emitter.cc [hunk 1]
```diff
@@ -130,13 +131,40 @@ ParallelLoopEmitter::EmitIndexAndSetExitBasicBlock(absl::string_view loop_name,
                       /*HasNUW=*/true, /*HasNSW=*/true);
   }
 
-  array_indices.emplace_back(linear_index_base, shape_, b_);
-  for (int i = 1; i < unroll_factor_; ++i) {
+  // When enable_row_index is true, it means the inner most dimensions
+  // match the block sizes.  So we can generate a simpler indexing
+  // for that dimensions.  This helps LLVM generate vectorized codes
+  // in that cases.
+  LaunchDimensions::Dim3D dim3 = launch_dimensions_.thread_counts_per_block();
```
Comment (akuegel) [inline tensorflow/compiler/xla/service/gpu/parallel_loop_emitter.cc:138]: It seems this variable is unused

**Review by akuegel — CHANGES_REQUESTED** [review]
Internally there were a lot of failing tests with this PR.
Can you please check whether you can reproduce the failures for these tests:
tensorflow/compiler/xla/service/gpu/tests:gemm_rewrite_test
tensorflow/compiler/xla/tests:array_elementwise_ops_test_gpu
tensorflow/compiler/tests:fused_batchnorm_test_gpu
tensorflow/compiler/tests:binary_ops_test_gpu

(there were more failures, but I hope at least one of those tests can be used to reproduce the problem).

**Comment by nouiz** [comment]
> Internally there were a lot of failing tests with this PR.
> Can you please check whether you can reproduce the failures for these tests:
> tensorflow/compiler/xla/service/gpu/tests:gemm_rewrite_test
> tensorflow/compiler/xla/tests:array_elementwise_ops_test_gpu
> tensorflow/compiler/tests:fused_batchnorm_test_gpu
> tensorflow/compiler/tests:binary_ops_test_gpu
> 
> (there were more failures, but I hope at least one of those tests can be used to reproduce the problem).

The first 3 are failing here. So I can repro it and work on it. Thanks.

I amended the last commit to remove the not used variable.

**Comment by nouiz** [comment]
I was able to repro tests failure. But I also had those same tests on the commits my PR was based on.
So I rebased and now only one tests fail. But it also fail on the commits my PR is based on.
It is a link failure. So this could be an environment issues.

Can you run again the CI?

**Review by akuegel — APPROVED** [review]

**Comment by akuegel** [comment]
> I was able to repro tests failure. But I also had those same tests on the commits my PR was based on.
> So I rebased and now only one tests fail. But it also fail on the commits my PR is based on.
> It is a link failure. So this could be an environment issues.
> 
> Can you run again the CI?

I approved the PR which made the tests run again. Unfortunately I still see a few failing tests, e.g. //tensorflow/compiler/tests:fused_batchnorm_test_gpu
Can you also see the results in the Details section of "Linux GPU" ? If not, I can give you the full list of failing tests.

**Comment by nouiz** [comment]
I fixed the failures that I can repro. Can you review the new commit and approuve?
I'm not able to run all tests locally. See bellow.

I'm able to see the public CI failures.

I ran on my PR this tests locally and it fails: //tensorflow/compiler/xla/tests:broadcast_test_gpu
But if I takes the base TF commit, it pass. So now I have a real failure from this PR.

With it, I was able to do one fix.

I tried to run all the tests, but I have tons of failures similar to this one:

```
ImportError: /root/.cache/bazel/_bazel_root/45a7ca8b5c99c684c2d5c22cdd8175f0/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/tests/jit_test_gpu.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN10tensorflow7functor12UnaryFunctorIN5Eigen9GpuDeviceENS0_3negINS2_4halfEEEEclERKS3_NS2_9TensorMapINS2_6TensorIS5_Li1ELi1ElEELi16ENS2_11MakePointerEEENSA_INSB_IKS5_Li1ELi1ElEELi16ESD_EE
```

So I can't test those and will need to rely on the CI...

I tried to use the google tensorflow/tensorflow:devel-gpu container. But I'm not able to even run the tensorflow configure scripts. It fail with this error:

```
ERROR: An error occurred during the fetch of repository 'local_config_cuda':
   Traceback (most recent call last):
        File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 1396, column 38, in _cuda_autoconf_impl
                _create_local_cuda_repository(repository_ctx)
        File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 977, column 35, in _create_local_cuda_repository
                cuda_config = _get_cuda_config(repository_ctx, find_cuda_config_script)
        File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 666, column 30, in _get_cuda_config
                config = find_cuda_config(repository_ctx, find_cuda_config_script, ["cuda", "cudnn"])
        File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 643, column 41, in find_cuda_config
                exec_result = _exec_find_cuda_config(repository_ctx, script_path, cuda_libraries)
        File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 637, column 19, in _exec_find_cuda_config
                return execute(repository_ctx, [python_bin, "-c", decompress_and_execute_cmd])
        File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/remote_config/common.bzl", line 230, column 13, in execute
                fail(
Error in fail: Repository command failed
Could not find any cublas_api.h matching version '' in any subdirectory:
        ''
        'include'
        'include/cuda'
        'include/*-linux-gnu'
        'extras/CUPTI/include'
        'include/cuda/CUPTI'
of:
        '/lib/x86_64-linux-gnu'
        '/usr'
        '/usr/lib/x86_64-linux-gnu'
        '/usr/local/cuda'
        '/usr/local/cuda-11.0/targets/x86_64-linux/lib'
        '/usr/local/cuda/lib64/stubs'
```

I wasn't able to make manually a working .tf_configure.bazelrc. CUDA and CUBLAS detection doesn't work anymore in that container. So I can't even run what the CI is running.

If you have any idea how to fix one or both of those problem, it would be great.

**Comment by nouiz** [comment]
I rebased as there was a conflict.

**Review by akuegel — APPROVED** [review]
Thanks, it seems now the tests pass :)
I saw a typo in the new function name, but I will fix it myself when merging.

**Comment by akuegel** [comment]
> I fixed the failures that I can repro. Can you review the new commit and approuve?
> I'm not able to run all tests locally. See bellow.
> 
> I'm able to see the public CI failures.
> 
> I ran on my PR this tests locally and it fails: //tensorflow/compiler/xla/tests:broadcast_test_gpu
> But if I takes the base TF commit, it pass. So now I have a real failure from this PR.
> 
> With it, I was able to do one fix.
> 
> I tried to run all the tests, but I have tons of failures similar to this one:
> 
> ```
> ImportError: /root/.cache/bazel/_bazel_root/45a7ca8b5c99c684c2d5c22cdd8175f0/execroot/org_tensorflow/bazel-out/k8-opt/bin/tensorflow/compiler/tests/jit_test_gpu.runfiles/org_tensorflow/tensorflow/python/_pywrap_tensorflow_internal.so: undefined symbol: _ZN10tensorflow7functor12UnaryFunctorIN5Eigen9GpuDeviceENS0_3negINS2_4halfEEEEclERKS3_NS2_9TensorMapINS2_6TensorIS5_Li1ELi1ElEELi16ENS2_11MakePointerEEENSA_INSB_IKS5_Li1ELi1ElEELi16ESD_EE
> ```
> 
> So I can't test those and will need to rely on the CI...
> 
> I tried to use the google tensorflow/tensorflow:devel-gpu container. But I'm not able to even run the tensorflow configure scripts. It fail with this error:
> 
> ```
> ERROR: An error occurred during the fetch of repository 'local_config_cuda':
>    Traceback (most recent call last):
>         File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 1396, column 38, in _cuda_autoconf_impl
>                 _create_local_cuda_repository(repository_ctx)
>         File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 977, column 35, in _create_local_cuda_repository
>                 cuda_config = _get_cuda_config(repository_ctx, find_cuda_config_script)
>         File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 666, column 30, in _get_cuda_config
>                 config = find_cuda_config(repository_ctx, find_cuda_config_script, ["cuda", "cudnn"])
>         File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 643, column 41, in find_cuda_config
>                 exec_result = _exec_find_cuda_config(repository_ctx, script_path, cuda_libraries)
>         File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/gpus/cuda_configure.bzl", line 637, column 19, in _exec_find_cuda_config
>                 return execute(repository_ctx, [python_bin, "-c", decompress_and_execute_cmd])
>         File "/home/fbastien/github/tensorflow-tf2-upstream2.cp/third_party/remote_config/common.bzl", line 230, column 13, in execute
>                 fail(
> Error in fail: Repository command failed
> Could not find any cublas_api.h matching version '' in any subdirectory:
>         ''
>         'include'
>         'include/cuda'
>         'include/*-linux-gnu'
>         'extras/CUPTI/include'
>         'include/cuda/CUPTI'
> of:
>         '/lib/x86_64-linux-gnu'
>         '/usr'
>         '/usr/lib/x86_64-linux-gnu'
>         '/usr/local/cuda'
>         '/usr/local/cuda-11.0/targets/x86_64-linux/lib'
>         '/usr/local/cuda/lib64/stubs'
> ```
> 
> I wasn't able to make manually a working .tf_configure.bazelrc. CUDA and CUBLAS detection doesn't work anymore in that container. So I can't even run what the CI is running.
> 
> If you have any idea how to fix one or both of those problem, it would be great.

Sorry to hear about the trouble with the open source build. Unfortunately I am not so familiar with the infrastructure stuff. I think I have heard that find_cuda_config is deprecated, and possibly the tensorflow/tensorflow:devel-gpu docker image needs to be updated to reflect that.
@chsigg Does that error look familiar?

**Comment by akuegel** [comment]
This has been merged now.

**Comment by nouiz** [comment]
Thanks.

**Comment by nouiz** [comment]
About the container problem. I used latest-gpu by mistake instead of devel-gpu. When I use devel-gpu, I do not have the error I reported. So this wasn't a real problem.
I still have the undefined symbol in my own container. But this is something else.

**Comment by akuegel** [comment]
Unfortunately the PR was automatically rolled back in d73d688687e48a29f06c84404f9817876e4b6325 due to two test failures:
//tensorflow/compiler/xla/tests:call_test_gpu
//tensorflow/compiler/xla/service:dynamic_padder_test_gpu
But the fix seems simple, ThreadsPerBlockRowVectorized should return -1 if the shape is empty. I will try to roll forward with the fix now.

**Comment by nouiz** [comment]
Thanks for looking at it. Why those tests fail, but we do not see any failure in the "Linux GPU" CI?

**Comment by akuegel** [comment]
> Thanks for looking at it. Why those tests fail, but we do not see any failure in the "Linux GPU" CI?

They only fail with --copt=-UNDEBUG (essentially an assert was triggered, but asserts are disabled by default with -c opt). But I think technically it could result in a segfault if unlucky.

**Comment by nouiz** [comment]
Thanks

**Pull request closed without merging**